src/combinatorics/set_family/compression/uv.lean

```diff
@@ -0,0 +1,237 @@
+  { refl }
+end
+
+@[simp] lemma compression_self (u : α) (s : finset α) : 𝓒 u u s = s :=
+begin
+  unfold compression,
+  convert union_empty s,
+  { ext a,
+    rw [mem_filter, compress_self, and_self] },
+  { refine eq_empty_of_forall_not_mem (λ a ha, _),
+    simp_rw [mem_filter, mem_image, compress_self] at ha,
+    obtain ⟨⟨b, hb, rfl⟩, hb'⟩ := ha,
+    exact hb' hb }
+end
+
+/-- Any family is compressed along two identical elements. -/
+lemma is_compressed_self (u : α) (s : finset α) : is_compressed u u s := compression_self u s
+
+lemma compress_disjoint (u v : α) :
+  disjoint (s.filter (λ a, compress u v a ∈ s)) ((s.image $ compress u v).filter (λ a, a ∉ s)) :=
+disjoint_left.2 $ λ a ha₁ ha₂, (mem_filter.1 ha₂).2 (mem_filter.1 ha₁).1
+
+/-- Compressing an element is idempotent. -/
+@[simp] lemma compress_idem (u v a : α) : compress u v (compress u v a) = compress u v a :=
+begin
+  unfold compress,
+  split_ifs with h h',
+  { suffices : u = ⊥,
+    { rw [this, sup_bot_eq, sup_bot_eq, _root_.sdiff_idem] },
+    rw ←disjoint_self,
+    have : u \ v = u := (h.1.mono_right h.2).sdiff_eq_left,
+    nth_rewrite 1 ←this,
+    exact h'.1.mono_right (sdiff_le_sdiff_right le_sup_right) },
+  { refl },
+  { refl }
+end
+
+lemma compress_mem_compression (ha : a ∈ s) : compress u v a ∈ 𝓒 u v s :=
+begin
+  rw mem_compression,
+  by_cases compress u v a ∈ s,
+  { rw compress_idem,
+    exact or.inl ⟨h, h⟩ },
+  { exact or.inr ⟨h, a, ha, rfl⟩ }
+end
+
+-- This is a special case of `compress_mem_compression` once we have `compression_idem`.
+lemma compress_mem_compression_of_mem_compression (ha : a ∈ 𝓒 u v s) : compress u v a ∈ 𝓒 u v s :=
+begin
+  rw mem_compression at ⊢ ha,
+  simp only [compress_idem, exists_prop],
+  obtain ⟨_, ha⟩ | ⟨_, b, hb, rfl⟩ := ha,
+  { exact or.inl ⟨ha, ha⟩ },
+  { exact or.inr ⟨by rwa compress_idem, b, hb, (compress_idem _ _ _).symm⟩ }
+end
+
+/-- Compressing a family is idempotent. -/
+@[simp] lemma compression_idem (u v : α) (s : finset α) : 𝓒 u v (𝓒 u v s) = 𝓒 u v s :=
+begin
+  have h : filter (λ a, compress u v a ∉ 𝓒 u v s) (𝓒 u v s) = ∅ :=
+    filter_false_of_mem (λ a ha h, h $ compress_mem_compression_of_mem_compression ha),
+  rw [compression, image_filter, h, image_empty, ←h],
+  exact filter_union_filter_neg_eq _ (compression u v s),
+end
+
+/-- Compressing a family doesn't change its size. -/
+lemma card_compression (u v : α) (s : finset α) : (𝓒 u v s).card = s.card :=
+begin
+  rw [compression, card_disjoint_union (compress_disjoint _ _), image_filter, card_image_of_inj_on,
+    ←card_disjoint_union, filter_union_filter_neg_eq],
+  { rw disjoint_iff_inter_eq_empty,
+    exact filter_inter_filter_neg_eq _ _ },
+  intros a ha b hb hab,
+  dsimp at hab,
+  rw [mem_coe, mem_filter, function.comp_app] at ha hb,
+  rw compress at ha hab,
+  split_ifs at ha hab with has,
+  { rw compress at hb hab,
+    split_ifs at hb hab with hbs,
+    { have h : (a ⊔ u) \ v \ u ⊔ v = (b ⊔ u) \ v \ u ⊔ v,
```

Comment:
Yaël and I spoke a bit about this in person but I'll add a brief comment here anyway so it doesn't get forgotten: this `have` gives a result which I originally had as a (weirdly stated) lemma in the original branch. The result is also useful in a later proof, so I think it should be expressed as a lemma; specifically a statement that `(- ⊔ u) \ v` is injective on particular sets (of course using `set.inj_on`)